orderbook/src/api/get_fee_and_quote.rs

```diff
@@ -0,0 +1,238 @@
+use crate::fee::{MinFeeCalculating, MinFeeCalculationError};
+use anyhow::Result;
+use chrono::{DateTime, Utc};
+use ethcontract::{H160, U256};
+use model::h160_hexadecimal;
+use model::{order::OrderKind, u256_decimal};
+use serde::{Deserialize, Serialize};
+use shared::{
+    conversions::{big_int_to_u256, U256Ext},
+    price_estimate::{PriceEstimating, PriceEstimationError},
+};
+use std::convert::Infallible;
+use std::sync::Arc;
+use warp::{hyper::StatusCode, reply, Filter, Rejection, Reply};
+
+#[derive(Deserialize)]
+struct Query {
+    #[serde(with = "h160_hexadecimal")]
+    sell_token: H160,
+    #[serde(with = "h160_hexadecimal")]
+    buy_token: H160,
+    // For sell orders in sell token, for buy orders in the buy token.
+    #[serde(with = "u256_decimal")]
+    in_amount_before_fees: U256,
+    kind: OrderKind,
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+struct Response {
+    expiration_date: DateTime<Utc>,
+    #[serde(with = "u256_decimal")]
+    min_fee: U256,
+    // For sell orders in buy token, for buy orders in the sell token.
+    #[serde(with = "u256_decimal")]
+    out_amount_before_fees: U256,
+    // For buy orders this is the same as out_amount_before_fees. For sell orders this is less
+    // than out_amount_before_fees because the fee is first deducted from the sell amount before
+    // applying the price.
+    #[serde(with = "u256_decimal")]
+    out_amount_after_fees: U256,
+}
+
+#[derive(Debug)]
+enum Error {
+    NotFound,
+    UnsupportedToken(H160),
+    AmountIsZero,
+    SellAmountDoesNotCoverFee,
+    Other(anyhow::Error),
+}
+
+impl From<MinFeeCalculationError> for Error {
+    fn from(other: MinFeeCalculationError) -> Self {
+        match other {
+            MinFeeCalculationError::NotFound => Error::NotFound,
+            MinFeeCalculationError::UnsupportedToken(token) => Error::UnsupportedToken(token),
+            MinFeeCalculationError::Other(error) => Error::Other(error),
+        }
+    }
+}
+
+impl From<PriceEstimationError> for Error {
+    fn from(other: PriceEstimationError) -> Self {
+        match other {
+            PriceEstimationError::UnsupportedToken(token) => Error::UnsupportedToken(token),
+            PriceEstimationError::Other(error) => Error::Other(error),
+        }
+    }
+}
+
+fn request() -> impl Filter<Extract = (Query,), Error = Rejection> + Clone {
+    warp::path!("fee_and_price")
```

Comment:
Is this the path? 

I would literally use this one: 
![image](https://user-images.githubusercontent.com/2352112/122577032-9f185600-d052-11eb-9005-31b28c893dfd.png)
https://protocol-rinkeby.dev.gnosisdev.com/api/#/default/get_api_v1_markets__baseToken___quoteToken___kind___amount_

It just changes the signature. But to make a gradual change, we can just create `/markets/next/...`  and after we migrate the frontends (balancer and cowswap) we kill the other two endpoints and we leave only the markers one. 

Would that make sense?

Reply:
I don't have a strong opinion on the path but I like the new one a little more. The bigger change with this route is that it doesn't put the parameters in the path like the markets route. Instead it puts them in the query like the fee route. Do you have a preference there?

Reply:
I found `/markets/<base>-<quote>/sell/<amount>` and `/markets/<base>-<quote>/buy/<amount>` very declarative, but maybe is a personal preference. 

I usually like APIs to be hierarchical, like `/markets/<base>-<quote>` modeling a market, and then with the these endpoints we get a sell/buy quote. The only difference is that now we return the fee together with the quote, which imo makes sense to be part of it. 

I let other give their opinions. I won't be opposed to the new name if that is the general sentiment.